ct: removing CT entries when policy is deleted

aanm opened the pull request:
When a policy is completely removed, the CT entries should also be clean
as it doesn't make sense to have CT if our policy behavior is default
deny. This PR removes the CT entries when a new policy added no longer
allows both endpoints to talk with each other.

Fixes https://github.com/cilium/cilium/issues/839

Notes: Recheck `PolicyDel` since it's being used when a kubernetes network policy is being added.

aanm commented on daemon/policy.go:
```diff
@@ -273,8 +273,15 @@ func (d *Daemon) PolicyDelete(labels labels.LabelArray) *apierror.APIError {
+		}
+	} else {
+		// TODO run GC only for endpoints that have the policy changed.
+		if d.PolicyEnabled() && d.policy.NumberOfRules() == 0 {
```
@ianvernon you might have to change this line for your work with #788

ianvernon replied:
Thanks for that! Duly noted.

ianvernon commented on daemon/ct.go:
```diff
@@ -103,3 +103,26 @@ func (d *Daemon) EnableConntrackGC() {
+			continue
+		}
+		e.Mutex.RUnlock()
+		// We can unlock the endpoint mutex sense
```
Nit: sense -> since

tgraf commented on daemon/ct.go:
```diff
@@ -103,3 +103,26 @@ func (d *Daemon) EnableConntrackGC() {
+		e.Mutex.RUnlock()
+		// We can unlock the endpoint mutex sense
+		// in runGC it will be locked as needed.
+		runGC(e, true, 0xFFFF) // 0xFFFF is the max interval possible.
```
Do I understand this correctly? You are expiring *all* entries of the conntrack table whenever policy is deleted. That is definitely not what we want. We want to expire CT entries which are related to policy rules which have been deleted.

aanm committed:
ct: add aliend_id in ct maps

Introducing the alien_id field to store the security label IDs of the
endpoint that created the CT entry.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
CT: cleaning entries when new policy denies them

Introduced the concept of GCFilter where each filter is in charge of
applying its logic to all entries in the CT map.

When a policy is removed from the repository, the global policy is
regenerated. If 2 endpoints are no longer allowed by the new policy
regenerated, the CT entries related to those 2 endpoints that are no
longer allowed to speak with each other and with this commit they
are removed.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/policy.go:
```diff
@@ -276,7 +277,41 @@ func (d *Daemon) PolicyDelete(labels labels.LabelArray) *apierror.APIError {
+			// FIXME: Do we really need to check the daemon.PolicyEnforcement()?
+			if ep.SecLabel == nil ||
+				!ep.Opts.IsEnabled(endpoint.OptionPolicy) ||
+				d.PolicyEnforcement() == endpoint.NeverEnforce {
```
Checking for `d.PolicyEnforcement()` seems useless and not required here.

ianvernon replied:
I agree, I do not see a reason to check for d.PolicyEnforcement here.

tgraf commented on daemon/policy.go:
```diff
@@ -276,7 +277,41 @@ func (d *Daemon) PolicyDelete(labels labels.LabelArray) *apierror.APIError {
+				}
+			}
+			log.Debugf("Keeping entries of EP %d: %+v", ep.ID, idsToKeep)
+			endpointmanager.KeepCTEntryOf(!d.conf.IPv4Disabled, ep, idsToKeep)
```
A new sec ID can appear while we construct this table and update it? In which case we might miss a new security identity popping up. No? Do we need to hold a lock to avoid updating consumables in the meantime or is it guaranteed that we pick up such changes later?

aanm replied:
Let me get this straight:

1. policy deleted
    1. waiting for consumables map to be regenerated
    2. retrieve a copy of consumables and generate the `IDsToKeep`
    3. start cleaning up all CT entries except `IDsToKeep`

2. a new (allowed by the current/new policy) consumer starts talking with the consumable after 1.ii

3. a new (denied by the new policy) consumer starts talking with the consumable after 1.ii

- [x] the endpoint created in 3 will be denied by the policy so there won't even be a CT entry.

- [ ] the endpoint created in 2 will be allowed by the policy but the CT entry will be removed because it's not in `IDsToKeep` - > This is the problem that you are seeing, right?

Ok, what do you suggest?

I'm thinking on holding the endpoint lock until all CT entries are cleaned.

This guarantees the endpoint won't get its `allowed-consumers` updated until its CT entries are cleaned up. But this also means I will need to move the `d.consumableCache.GetConsumables()` inside the for loop and get the `IDsToKeep` for each endpoint causing different endpoints (with the same security ID) to get different `IDsToKeep` map. :-(

tgraf replied:
In your example, I'm not worried about new CT entries. Your case "2" has to depend on an existing allow consumer so it will be fine.

What I'm worried about is a new security identity popping up while we generate the `IDsToKeep` list. The new security identity is properly allowed in the policymap and a TCP connection is established. This will then get dropped.

Could we instead create the list of identities which we have to *remove*? This would make this race condition impossible.

ianvernon commented on tests/03-docker.sh:
```diff
@@ -64,4 +64,12 @@ ping6 -c 5 "$SERVER_IP" || {
+ping6 -c 2 "$SERVER_IP" && {
+	abort "Error: Unexpected connectivity between host and server after policy removed"
+}
+
```
I like that you added checks for this throughout the other tests.

ianvernon commented on Jenkinsfile:
```diff
@@ -22,8 +22,6 @@ pipeline {
     post {
         always {
-            sh './tests/copy_files'
-            archiveArtifacts "cilium-files.tar.gz"
```
This change should be reverted.

ianvernon reviewed (approved):
Please address my one comment - the rest looks good to me.
